src/extended_debuggability.js

```diff
@@ -0,0 +1,304 @@
+        Promise.onChainLengthLimitExceeded =
+            function (fn) {chainLengthLimitHandler = fn;};
+    }
+}
+
+function disableChainLengthLimit() {
+    if (chainLengthLimit > 0) {
+        util.unhookFrom(Promise.prototype, "_hook_chained",
+            incrementChainLength);
+        chainLengthLimitHandler = exceptionChainLengthLimitHandler;
+        chainLengthLimit = null;
+    }
+}
+
+// Pending promises limit
+var pendingPromisesLimit = null;
+var pendingPromisesNumber = 0;
+var exceptionPendingPromisesLimitHandler = function() {
+    throw new Error("Too many pending promises, it reached limit" +
+        " of " + pendingPromisesLimit + " promises");
+};
+var pendingPromisesLimitHandler = exceptionPendingPromisesLimitHandler;
+
+function incrementPending() {
+    if (pendingPromisesLimit > 0) {
+        pendingPromisesNumber++;
+        if (pendingPromisesNumber > pendingPromisesLimit) {
+            pendingPromisesLimitHandler();
+        }
+    }
+}
+
+function decrementPending() {
+    if (pendingPromisesLimit)
+        pendingPromisesNumber--;
+}
+
+function enableMaxPendingPromises(limit) {
+    if (pendingPromisesLimit) return;
+    pendingPromisesLimit = limit;
+    util.hookTo(Promise.prototype, "_hook_created", incrementPending);
+    util.hookTo(Promise.prototype, "_hook_fulfilled", decrementPending);
+    util.hookTo(Promise.prototype, "_hook_rejected", decrementPending);
+    util.hookTo(Promise.prototype, "_hook_following", decrementPending);
+    util.hookTo(Promise.prototype, "_hook_cancelled", decrementPending);
+    Promise.onPendingPromisesLimitExceeded =
+            function (fn) {pendingPromisesLimitHandler = fn;};
+}
+
+function disableMaxPendingPromises() {
+    if (!pendingPromisesLimit) return;
+    pendingPromisesLimit = null;
+    pendingPromisesNumber = 0;
+    pendingPromisesLimitHandler = exceptionPendingPromisesLimitHandler;
+    util.unhookFrom(Promise.prototype, "_hook_created", incrementPending);
+    util.unhookFrom(Promise.prototype, "_hook_fulfilled", decrementPending);
+    util.unhookFrom(Promise.prototype, "_hook_rejected", decrementPending);
+    util.unhookFrom(Promise.prototype, "_hook_following", decrementPending);
+    util.unhookFrom(Promise.prototype, "_hook_cancelled", decrementPending);
+}
+
+// Promises tracing
+var tracingEnabled = false;
+var promiseTraceIdCounter = 0;
+
+function addTracingInfo (next) {
+    if (!this._promiseId) {
+        promiseTraceIdCounter++;
+        this._promiseId = promiseTraceIdCounter;
+    }
+    if (!next._promiseId) {
+        promiseTraceIdCounter++;
+        next._promiseId = promiseTraceIdCounter;
+    }
+    if (!this._tracks) this._tracks = [];
+    if (!next._trackedBy) next._trackedBy = [];
+    this._tracks.push(next);
+    next._trackedBy.push(this);
+}
+
+// Recursive depth of this function is limited so it would produce
+// human-readable input and would not care about circles
+function getDOTGraph (depth, destination) {
```

Comment:
This stuff is to be implemented by users, the monitor api should only expose low level events and nothing else.